### ISSUE TITLE
Early return operator for Result

### ISSUE THREAD

**Issue by sporto** [gleam-lang/suggestions#27]
Just like `?` in Rust, this make the code a lot easier to understand when you need to deal with multiple results / options.

From

```
get_one(...)
   |> result.then(_, fn(one) {
      get_two(...)
         |> result.map(_, fn(two) {
             use_one_and_two(one, two)
         })
})
```

To

```
let one = get_one(...)?
let two = get_two(...)?
Ok(use_one_and_two(one, two))
```

**Comment by lpil** [comment]
Gleam doesn't currently have traits so `?` would be a lot more restrictive than in Rust (no into trait).

It would be good to also investigate OCaml's rebindable syntax which is used to achieve this and more.

**Comment by OvermindDL1** [comment]
If you add an implicit modules system then you have your trait-like things (though even more significantly powerful if first-class functors are added), then you'd be able to have something desugar as you want.

However, `?` fits in languages like Rust fine, because early returns, however I wouldn't recommend it for something like Gleam, which doesn't.  OCaml *already* has a construct for this.

OCaml lets you add an unbound amount of `+`'s or `*`'s at the end of a `let` or `and`, so you can do this:
```ocaml
let open Result in
let+ one = get_one(...) in
let+ two = get_two(...) in
...
```
And this just desugars to:
```ocaml
let open Result in
(let+) get_one(...) @@ fn one ->
(let+) get_two(...) @@ fn two ->
...
```
The `let+` in this case is defined in the `Result` module like:
```ocaml
let (let+) res monad =
  match res with
  | Err _ = e -> e
  | Some s -> monad s
```

But in general the `let* <match> = <expr> in <body>` gets mapped to `(let*) <expr> @@ fn <match> -> <body>`

This ability let's you open a module into the local space as normal, then use them like `let`'s.  On the forums I built a whole Elixir-style `for` comprehension with them and described these in more detail especially how they work in regards to the whole patterns.

In General, for consistency, the OCaml community uses `+`'s as meaning a monad call (where it takes something, operates on the innerds, and returns that same type), and where `*`'s mean an applicative call.  It is considered that they may open up other operators to be used on them as well, but already the OCaml community is making great use of these with not only monad/applicative style handling for things like Result handling, but also for comprehension systems, parsers, etc... etc...  It makes a *lot* of code a *lot* more readable, and it handles the `?` case just fine.

**Comment by lpil** [comment]
Some syntax ideas:

```rust
import gleam/result.{then}

pub fn main() -> Result(Thing, Err) {
  let#then value = may_fail()
  let#then next = also_may_fail(value)
  Ok(next)
}

pub fn qualified_main() -> Result(Thing, Err) {
  let#result.then value = may_fail()
  let#result.then next = also_may_fail(value)
  Ok(next)
}
```

Quite clear, but also quite verbose.

**Comment by lpil** [comment]
What other approaches do languages take here? 

- Haskell's do syntax
- Early returns
- Throw
- Exceptions

**Comment by c-cube** [comment]
something that generalizes exceptions pretty well, in a typed setting, is algebraic effects.

See [koka](https://github.com/koka-lang/koka) (and [slides](http://research.microsoft.com/en-us/projects/koka/2012-overviewkoka.pdf)), the [plans for OCaml](http://ocamllabs.io/doc/effects.html), etc.

**Comment by CrowdHailer** [comment]
My opinion is you don't need something that generalises, or that you can discover the generalisation in later versions.

If result is going to get special status (as the comment on this issue suggests to me, https://github.com/gleam-lang/gleam/issues/531) Then you could have something that worked just for results.

I like the rust approach, and even without traits It would be useful. 
The only thing I don't like about the rust approach is `?` very much suggests boolean to me. I would have chose `!` instead.

I also think a syntax like below would work, 
```
let value != try_get_value()
```

**Comment by lpil** [comment]
To me that reads as "let value not equal to try_get_value".

If we specialise for Result I think I would prefer this syntax:

```rust
try value = try_get_value()
```

**Comment by lpil** [comment]
@tomwhatmore @gamebox @QuinnWilton @chouzar @kyle-sammons Do you have any thoughts on the above?

**Comment by gamebox** [comment]
I've let it be known before that I think a comprehension is as readable and more powerful than one off syntax like this.  It can be specialized to the built-in monad-likes we have (List and Result), and can be generalized with whatever polymorphism device we develop later(traits, first class modules, etc).

**Comment by tomwhatmore** [comment]
I'm a big fan of `?` in Rust and would very much like such a thing to exist in Gleam. As to the specifics...

* My entire knowledge of OCaml is a couple of afternoons playing around with reasonml so not remotely qualified to judge its implementation.
* I think I would agree that it probably doesn't need a general approach, I think specialising to Result (and thus Option as well) is fine. I have personally never used `?` with anything other than the default types in Rust and still find it extremely useful.
* `let value != try_get_value()` reads as not equal to me also.
* I think Rust has a number of weird naming decisions round results (`.expect()` for example) but `?` doesn't seem bad to me. I would be fine with `!` as well, but at the end rather than by the `=`.
* try looks like Swift (though it would be `let value = try get_value()` over there) and would work quite similarly to how you'd use it in a Swift function that is marked with throws. I quite like it, but this may be because I have written a lot of swift.
* The specialised form seems like it adds immediate value to the existing fundamental components of Gleam. Trying to work out a generalised form that would in theory work nicely with hypothetical parts of gleam that could exist in some form in the future sounds much harder. Perhaps I just lack ambition.

This seems like an important point that didn't get any discussion:

> However, ? fits in languages like Rust fine, because early returns, however I wouldn't recommend it for something like Gleam, which doesn't.

Is there a specific rationale for not having early returns in Gleam? Would supporting them have any other ramifications?

**Comment by lpil** [comment]
> Is there a specific rationale for not having early returns in Gleam? Would supporting them have any other ramifications?

I think that early returns are a lot more useful in statement based languages than expression based languages as they don't pllay well with higher order functions. I can't say I've ever wanted them in Erlang/Elixir/Haskell etc, so I don't feel a reason to add them to Gleam at present.

**Comment by kyle-sammons** [comment]
Chiming a bit late here, but I agree with most things that @tomwhatmore said. I also prefer the syntax of `try value = may_fail_func()` as it seems much more clear-cut and readable. That said, as a Haskell fan I'm also always a fan of the `do` syntax, but that may be overkill for this specific use-case.

**Comment by lpil** [comment]
If we were to have a `try` binding syntax would it make more sense to have to work on patterns rather than to have it be specialised to Result?

```rust
// result based version
try form = read_form(request)

// expands to
case read_form(request) {
  Ok(form) -> ...
  Error(e) -> Error(e)
}
```

```rust
// pattern based version
try Ok(form) = read_form(request)

// expands to
case read_form(request) {
  Ok(form) -> ...
  Error(e) -> Error(e)
}
```

This would allow it to be used with types other than Result, and would match `assert` and `let` more, though it would be slightly more verbose.

**Comment by CrowdHailer** [comment]
I would lean to result based version, because other matches might not make as much sense, also you need a consistent type of the failure case for the early return.
The expanded code needs to return in the error case.

```rust
// result based version
try form = read_form(request)

// expands to
case read_form(request) {
  Ok(form) -> ...
  Error(e) -> return Error(e)
}
```

*I guess you example might actually already include this if you assume the `...` on the Ok branch includes all the lines of code after `try` in the function.*

**Comment by lpil** [comment]
Yes, all the lines after the try statement go in place of `...` as there is no return statement in Gleam.

**Comment by lpil** [comment]
I'm trying out the Result specialised `try` here https://github.com/gleam-lang/stdlib/commit/4787386533f909a46d9f9854af846a2d398d2586
